test/functional/options/defaults_spec.lua

```diff
@@ -328,3 +332,235 @@ describe('XDG-based defaults', function()
+    end)
+
+    it("doesn't expand $VARIABLES", function()
+      clear({env={
+        XDG_DATA_HOME='$VARIABLES',
+        VARIABLES='this-should-not-happen'
+      }})
+      eq(alter_path('$VARIABLES' .. appended_dir), funcs.stdpath('data'))
+    end)
+
+    it("doesn't expand ~/", function()
+      clear({env={
+        XDG_DATA_HOME='~/frobnitz'
+      }})
+      eq(alter_path('~/frobnitz' .. appended_dir), funcs.stdpath('data'))
+    end)
+  end)
+
+  describe('"cache"' , function ()
+    it('knows XDG_CACHE_HOME', function()
+      clear({env={
+        XDG_CACHE_HOME=alter_path('/home/docwhat/.cache')
+      }})
+      eq(alter_path('/home/docwhat/.cache/nvim'), funcs.stdpath('cache'))
+    end)
+
+    it('handles changes during runtime', function()
+      clear({env={
+        XDG_CACHE_HOME=alter_path('/home/original')
+      }})
+      eq(alter_path('/home/original/nvim'), funcs.stdpath('cache'))
+      nvim('command', 'let $XDG_CACHE_HOME="/home/new"')
+      eq(alter_path('/home/new/nvim'), funcs.stdpath('cache'))
+    end)
+
+    it("doesn't expand $VARIABLES", function()
+      clear({env={
+        XDG_CACHE_HOME='$VARIABLES',
+        VARIABLES='this-should-not-happen'
+      }})
+      eq(alter_path('$VARIABLES/nvim'), funcs.stdpath('cache'))
+    end)
+
+    it("doesn't expand ~/", function()
+      clear({env={
+        XDG_CACHE_HOME='~/frobnitz'
+      }})
+      eq(alter_path('~/frobnitz/nvim'), funcs.stdpath('cache'))
+    end)
+  end)
+
+  describe('"config_dirs"' , function ()
+    it('knows XDG_CONFIG_DIRS', function()
+      clear({env={
+        XDG_CONFIG_DIRS=alter_path('/home/docwhat/.config:/etc/config')
+      }})
+      eq({
+        alter_path('/home/docwhat/.config/nvim'),
+        alter_path('/etc/config/nvim'),
+      }, funcs.stdpath('config_dirs'))
+    end)
+
+    it('handles changes during runtime', function()
+      clear({env={
+        XDG_CONFIG_DIRS=alter_path('/home/original:/etc/config')
+      }})
+      eq({
+        alter_path('/home/original/nvim'),
+        alter_path('/etc/config/nvim')
+      }, funcs.stdpath('config_dirs'))
+      nvim('command', 'let $XDG_CONFIG_DIRS="/home/new:/etc/config"')
+      eq({
+        alter_path('/home/new/nvim'),
+        alter_path('/etc/config/nvim')
+      }, funcs.stdpath('config_dirs'))
+    end)
+
+    it("doesn't expand $VAR and $IBLES", function()
+      clear({env={
+        XDG_CONFIG_DIRS='$VAR:$IBLES',
+        VAR='this-should-not-happen',
+        IBLES='neither-should-this'
+      }})
+      eq({
+        alter_path('$VAR/nvim'),
+        alter_path('$IBLES/nvim'),
+      }, funcs.stdpath('config_dirs'))
+    end)
+
+    it("doesn't expand ~/", function()
+      clear({env={
+        XDG_CONFIG_DIRS='~/.oldconfig:~/.olderconfig'
+      }})
+      eq({
+        alter_path('~/.oldconfig/nvim'),
+        alter_path('~/.olderconfig/nvim'),
+      }, funcs.stdpath('config_dirs'))
+    end)
+  end)
+
+  describe('"data_dirs"' , function ()
```

Comment:
By “Colon-separated variable contains no directories” I meant simply “it is empty”. About “larger then MAXPATHL”: best if you just return those large directories. Maybe omitting them from the output. But not truncating (not here, I believe omitting is better then truncating for `stdpath`) and definitely not crashing Neovim (there were such problems, and maybe are now in some places). Querying Neovim for this value is not possible currently, but I think it is good idea to add `GREAT_DIR_LENGTH` constant to `test.functional.helpers`, needs to be at least twice as large as MAXPATHL (if it is not too large Neovim may not crash on buffer overflow).